torch/utils/data/datapipes/iter/dataframes.py

```diff
@@ -0,0 +1,434 @@
+        for item in self.source_datapipe:
+            yield self.output_var.calculate_me(item)
+
+
+@functional_datapipe('dataframes_as_tuples')
+class DataFramesAsTuplesPipe(IterDataPipe):
+    def __init__(self, source_datapipe):
+        self.source_datapipe = source_datapipe
+
+    def __iter__(self):
+        for df in self.source_datapipe:
+            for record in df.to_records(index=False):
+                yield record
+
+
+@functional_datapipe('dataframes_per_row', is_df=True)
+class PerRowDataFramesPipe(DFIterDataPipe):
+    def __init__(self, source_datapipe):
+        self.source_datapipe = source_datapipe
+
+    def __iter__(self):
+        for df in self.source_datapipe:
+            for i in range(len(df.index)):
+                yield df[i:i + 1]
+
+
+@functional_datapipe('dataframes_concat', is_df=True)
+class ConcatDataFramesPipe(DFIterDataPipe):
+    def __init__(self, source_datapipe, batch=3):
+        self.source_datapipe = source_datapipe
+        self.batch = batch
+        if not WITH_PANDAS:
+            Exception('DataFrames prototype requires pandas to function')
+
+    def __iter__(self):
+        buffer = []
+        for df in self.source_datapipe:
+            buffer.append(df)
+            if len(buffer) == self.batch:
+                yield pandas.concat(buffer)
+                buffer = []
+        if len(buffer):
+            yield pandas.concat(buffer)
+
+
+@functional_datapipe('dataframes_shuffle', is_df=True)
+class ShuffleDataFramesPipe(DFIterDataPipe):
+    def __init__(self, source_datapipe):
+        self.source_datapipe = source_datapipe
+        if not WITH_PANDAS:
+            Exception('DataFrames prototype requires pandas to function')
+
+    def __iter__(self):
+        size = None
+        all_buffer = []
+        for df in self.source_datapipe:
+            if size is None:
+                size = len(df.index)
+            for i in range(len(df.index)):
+                all_buffer.append(df[i:i + 1])
+        random.shuffle(all_buffer)
+        buffer = []
+        for df in all_buffer:
+            buffer.append(df)
+            if len(buffer) == size:
+                yield pandas.concat(buffer)
+                buffer = []
+        if len(buffer):
+            yield pandas.concat(buffer)
+
+
+@functional_datapipe('dataframes_filter', is_df=True)
+class FilterDataFramesPipe(DFIterDataPipe):
+    def __init__(self, source_datapipe, filter_fn):
+        self.source_datapipe = source_datapipe
+        self.filter_fn = filter_fn
+        if not WITH_PANDAS:
+            Exception('DataFrames prototype requires pandas to function')
+
+    def __iter__(self):
+        size = None
+        all_buffer = []
+        filter_res = []
+        for df in self.source_datapipe:
+            if size is None:
+                size = len(df.index)
+            for i in range(len(df.index)):
+                all_buffer.append(df[i:i + 1])
+                filter_res.append(self.filter_fn(df.iloc[i]))
```

Comment:
it looks like `df.iloc[i]` returns the ith series (i.e. Column), while `df.iloc[[i]]` returns the ith row (confusing Pandas syntax, lol) . https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html